### PR TITLE
Update DBConnection

### LINKED ISSUE #235: Update DBConnection
As per elixir-ecto/db_connection#167 the db_connection package has been updated. That issues outlines the steps needed to update the package.

### PR DESCRIPTION
**Pull request by Swingcloud**
closes #235 

I've try to update to connection from 1.x to 2.0.1, but seems some test didn't pass
1. in test auth connection leak sometimes will have one tcp_count when test just started, thinking about the way to clear ports before this test start
2. in Elixir v 1.6 will got a warning that some db_connection behaviour didn't use
```
lib/mongo/protocol.ex:1: Undefined callback function handle_prepare/3 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_rollback/2 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_fetch/4 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_declare/4 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_deallocate/4 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_close/3 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_begin/2 (behaviour ''Elixir.DBConnection'')
lib/mongo/protocol.ex:1: Undefined callback function handle_commit/2 (behaviour ''Elixir.DBConnection'')
```

but I didn't see any behaviours before, just wonder how to bypass those warning

### CODE REVIEW

**Comment by ankhers** [comment]
Thanks for this. I will take a look through it shortly.

File: lib/mongo/protocol.ex [hunk 1]
```diff
@@ -138,24 +139,8 @@ defmodule Mongo.Protocol do
     end
   end
 
-  def handle_info({:tcp, data}, s) do
-    err = Mongo.Error.exception(message: "unexpected async recv: #{inspect data}")
-    {:disconnect, err, s}
-  end
```
Comment (ankhers) [inline lib/mongo/protocol.ex:144]: I haven't looked into the changest, but are these messages no longer being sent?

File: test/mongo/connection_test.exs [hunk 2]
```diff
@@ -94,10 +95,10 @@ defmodule Mongo.ConnectionTest do
...
             backoff_type: :stop, auth_source: "admin_test"]
 
-    capture_log fn ->
-      assert {:ok, pid} = Mongo.start_link(opts)
-      assert_receive {:EXIT, ^pid, {%Mongo.Error{code: 18}, _}}
-    end
+    assert capture_log(fn ->
+      {:ok, pid} = Mongo.start_link(opts)
+      assert_receive {:EXIT, ^pid, :killed}, 5000
+    end) =~ "(Mongo.Error) auth failed for user mongodb_admin_user"
```
Comment (ankhers) [inline test/mongo/connection_test.exs:101]: What was the purpose of changing the asserts?

File: lib/mongo/protocol.ex (hunk 1 above)
Comment (Swingcloud) [reply lib/mongo/protocol.ex:144]: seems latest dbconnect doesn't have the `handle_info` callback method, so I just deleted it, but I am not sure this is used by dbconnect or not

File: test/mongo/connection_test.exs (hunk 2 above)
Comment (Swingcloud) [reply test/mongo/connection_test.exs:101]: after I updated DBConnection, the old way of capturing log is no longer useful, so I use `ExUnit.CaptureLog` to make sure it did return some error message

**Comment by kiote** [comment]
hi @ankhers, do you need any help here? Really interested in this upgrade

**Comment by ankhers** [comment]
@kiote I think the easiest thing to do would be to just test this branch to make sure things are working correctly. Most important would be testing various failure scenarios (closed connections, that sort of thing).

**Comment by AlexKovalevych** [comment]
Any news on this one?

**Comment by kiote** [comment]
I ended up with some workaround / didn't have time to check this 🤷‍♀️

**Comment by ankhers** [comment]
Unfortunately I have not been able to test these changes yet either. I am hoping to find a bit of time soon.

**Comment by morganatwishpond** [comment]
I merged master locally and the entire test suite has passed. Will be testing this out in our staging environment shortly.

**Comment by foggy1** [comment]
If I run from this branch and just pop into an interactive session, sitting there briefly, the following happens: 

```
** (stop) exited in: GenServer.call(#PID<0.373.0>, {:server_description, %{address: "localhost:27017", arbiters: [], election_id: nil, error: nil, hosts: [], last_update_time: -576460660915, last_write_date: nil, max_wire_version: 5, me: nil, min_wire_version: 0, op_time: nil, passives: [], primary: nil, round_trip_time: 4, set_name: nil, set_version: nil, tag_set: %{}, type: :standalone}}, 30000)
    ** (EXIT) time out
```

If I try to make an explicit `find/4` call to the local Mongo server:

```
** (exit) exited in: GenServer.call(:mongo, :topology, 5000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:924: GenServer.call/3
    (mongodb) lib/mongo.ex:819: Mongo.select_servers/4
    (mongodb) lib/mongo.ex:796: Mongo.select_server/3
    (mongodb) lib/mongo.ex:442: Mongo.find/4
```
EDIT: have confirmed that raw `start_link` runs without pooling. Not sure what's up with supervisor and pooling timeouts, though.

**Comment by morganatwishpond** [comment]
Update: so far no luck. Updating this driver to use dbconnection 2.0 is beyond me. :(

> @kiote I think the easiest thing to do would be to just test this branch to make sure things are working correctly. Most important would be testing various failure scenarios (closed connections, that sort of thing).

@ankhers is right about this - one of the big problems I wasn't able to figure out (and I suspect it's due the callbacks not being implemented) is that connections just swell up and never get closed when using pooling.

@ankhers : Any chance you'll be able to take a swing at this PR?

**Comment by ankhers** [comment]
I will try to find the time to dig through this soon.

**Comment by ankhers** [comment]
I am closing this in favour of #279. Thank you for the contribution!

**Pull request closed without merging**